### ISSUE TITLE
1.12.2 Heap corruption on Windows

### ISSUE THREAD

**Issue by woytaz** [pocoproject/poco#3824]
Hey,

When trying to use POCO 1.12.0-1.12.2 on Windows (compiling with Visual Studio 2022 17.3, MSVC 14.33) I'm met with a heap corruption error:
<img src="https://user-images.githubusercontent.com/4399620/194323385-123f7845-6de0-4c47-b7de-e6af15158ea0.png" width=50% height=50%>

The minimal code that I managed to cause the corruption with is as follows:
```c++
#define POCO_NO_SOO
#include <Poco/JSON/Parser.h>

int main()
{
	auto handler{Poco::makeShared<Poco::JSON::ParseHandler>()};
}
```
(`POCO_NO_SOO` because otherwise I get an Internal Compiler Error in `Any`)

With the help of Conan, I prepared a simple project, which reproduces the issue: [Corruption.zip](https://github.com/pocoproject/poco/files/9725158/Corruption.zip)
With `conan test` you can easily test different versions of POCO against the above project, e.g. assuming you're in the project's directory:
```
conan test . poco/1.12.2 --build=missing
```
This will build POCO and its dependencies, based on the conan.io recipe: https://conan.io/center/poco
(Note: the test package's conanfile also specifies `*:shared=True` so all libraries will be using in their shared variant as that is my use case.)

The output should indicate no errors and the following should be printed:
```
Creating handler.
Handler created.
Handler destroyed.
```
Instead, only the first two messages appear in the output and then the application crashes (in release) or shows the above dialog (in debug).

For completeness I'll paste the Conan profile I'm using:
```ini
[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
compiler=Visual Studio
compiler.version=17
build_type=Debug
```
(`RelWithDebInfo` for the release build.)

I've also tried this on a different machine, using a custom Conan recipe, which downloads POCO's sources, uses unbundled dependencies, and compiles it all using VS 2019 toolchain (14.29). The same happens for the unreleased POCO 1.12.3, but _not_ for 1.11.3 or 1.10.1.

In a modified project, when I had a debugger attached, I got the following call stack: [link](https://user-images.githubusercontent.com/4399620/194330001-68a15f72-47fe-40b8-918b-ac7d75cfdf8d.png).

To sum up, the problem:
* Occurs in POCO versions 1.12.0-1.12.2, as well as the as-yet-unreleased 1.12.3.
* Occurs when using MSVC 14.29 (VS 2019) and 14.33 (VS 2022).
* Occurs on two different machines.
* Does NOT occur in POCO versions 1.10.1 and 1.11.3.

**Comment by aleks-f** [comment]
You should not define `POCO_NO_SOO` in the application - that define must be on the library level and libraries recompiled, otherwise (as you noticed) you may see funny things happen.

Furthermore, the compiler error you're seeing is a confirmed compiler bug - see #3682, #3683 and [fatal error C1001: Internal compiler error. Two Phase Look Up failed on Template Function](https://developercommunity.visualstudio.com/t/fatal-error-C1001:-Internal-compiler-err/10095155#T-N10099115).

Sadly, the fix did not make it in 1.12.2 but it will be in [1.12.3](https://github.com/pocoproject/poco/commit/1d10aeb248a930a004a6689b4fa4600b30ebff7f)
